action/protocol/staking/candidate_center.go

```diff
@@ -81,8 +81,14 @@ func (m CandidateCenter) GetBySelfStakingIndex(index uint64) *Candidate {
 	return nil
 }
 
-// Put writes the candidate into map
-func (m CandidateCenter) Put(d *Candidate) {
+// Upsert adds a candidate into map, overwrites if already exist
+func (m CandidateCenter) Upsert(d *Candidate) {
+	if c, ok := m.ownerMap[d.Owner.String()]; ok {
+		delete(m.nameMap, c.Name)
+		delete(m.operatorMap, c.Operator.String())
+		delete(m.selfStkBucketMap, c.SelfStakeBucketIdx)
+	}
```

Comment:
As nameMap/operatorMap/selfStkBucketMap are all indexers of d inside of CandidateCenter, if an upsert of d is intended, to maintain the consistency this struct, we are supposed to make sure (take name map for example):
m.nameMap[c.Name] is either empty or identical to d, if not, throw an error.

This sounds like a duplicate logic, but it may be better to keep the logic inside of this struct instead.
```
	if c, ok := m.ownerMap[d.Owner.String()]; ok {
		if nd, ok := m.nameMap[c.Name]; ok && nd.Owner.String() != d.Owner.String() {
                        return errors.New("duplicate name")
                 }
                 ... ...
	}

Reply:
I think m.nameMap[c.Name]  should always link to c, since only CandidateCenter itself can insert to all maps it own

Reply:
updated in latest push